src/base/ClpMap.h

```diff
@@ -0,0 +1,248 @@
+    void trim(size_t wantSpace = 0);
+    bool del(const KeyMapIterator &);
+    void findEntry(const Key &, KeyMapIterator &);
+    size_t memoryCountedFor(const Key &, const EntryValue *);
+
+    /// The {key, value, ttl} tuples.
+    /// Currently stored and maintained in LRU sequence.
+    Storage data;
+
+    /// index of stored data by key
+    KeyMapping index;
+
+    /// TTL to use if none provided to add(). 0 to disable caching.
+    int defaultTtl = std::numeric_limits<int>::max;
+    size_t memLimit_ = 0; ///< The maximum memory to use
+    size_t memUsed_ = 0;  ///< The amount of memory currently used
+    int entries_ = 0;     ///< The stored entries
+};
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+ClpMap<Key, EntryValue, MemoryUsedByEV>::ClpMap(int aTtl, size_t aSize) :
+    defaultTtl(aTtl)
+{
+    setMemLimit(aSize);
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+void
+ClpMap<Key, EntryValue, MemoryUsedByEV>::setMemLimit(size_t aSize)
+{
+    const auto oldLimit = memLimit_;
+    if (aSize > 0)
+        memLimit_ = aSize;
+    else
+        memLimit_ = 0;
+
+    if (oldLimit > memLimit_)
+        trim();
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+void
+ClpMap<Key, EntryValue, MemoryUsedByEV>::findEntry(const Key &key, KeyMapIterator &i)
+{
+    i = index.find(key);
+    if (i == index.end()) {
+        return;
+    }
+
+    auto &e = (*i).second;
+    if (!expired(*e)) {
+        if (e != data.begin())
+            data.splice(data.begin(), data, e, std::next(e));
+        return;
+    }
+    // else fall through to cleanup
+
+    del(i);
+    i = index.end();
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+EntryValue *
+ClpMap<Key, EntryValue, MemoryUsedByEV>::get(const Key &key)
+{
+    KeyMapIterator i;
+    findEntry(key, i);
+    if (i != index.end()) {
+        const Entry &e = *(i->second);
+        return e.value;
+    }
+    return NULL;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+size_t
+ClpMap<Key, EntryValue, MemoryUsedByEV>::memoryCountedFor(const Key &k, const EntryValue *v)
+{
+    // TODO: handle Entry which change size while stored
+    size_t entrySz = sizeof(Entry) + MemoryUsedByEV(v) + k.length();
+    return sizeof(MapItem) + k.length() + entrySz;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t)
+{
+    return add(key, t, defaultTtl);
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t, int ttl)
+{
+    if (ttl == 0)
+        return false;
+
+    if (memLimit() == 0)
+        return false;
+
+    del(key);
+
+    const auto wantSz = memoryCountedFor(key, t);
```

Comment:
I recommend renaming wantSz to wantSpace. Our trim() is already using that name; it clarifies the true meaning of this variable.

Reply:
Done in 9cd838d3c